src/hotspot/share/opto/ifnode.cpp

```diff
@@ -539,33 +539,38 @@ int RangeCheckNode::is_range_check(Node* &range, Node* &index, jint &offset) {
 
 //------------------------------adjust_check-----------------------------------
 // Adjust (widen) a prior range check
-static void adjust_check(Node* proj, Node* range, Node* index,
+static void adjust_check(IfProjNode* proj, Node* range, Node* index,
                          int flip, jint off_lo, PhaseIterGVN* igvn) {
   PhaseGVN *gvn = igvn;
   // Break apart the old check
   Node *iff = proj->in(0);
   Node *bol = iff->in(1);
   if( bol->is_top() ) return;   // In case a partially dead range check appears
   // bail (or bomb[ASSERT/DEBUG]) if NOT projection-->IfNode-->BoolNode
-  DEBUG_ONLY( if( !bol->is_Bool() ) { proj->dump(3); fatal("Expect projection-->IfNode-->BoolNode"); } )
-  if( !bol->is_Bool() ) return;
+  DEBUG_ONLY( if (!bol->is_Bool()) { proj->dump(3); fatal("Expect projection-->IfNode-->BoolNode"); } )
+  if (!bol->is_Bool()) return;
 
   Node *cmp = bol->in(1);
   // Compute a new check
   Node *new_add = gvn->intcon(off_lo);
-  if( index ) {
-    new_add = off_lo ? gvn->transform(new AddINode( index, new_add )) : index;
+  if (index) {
+    new_add = off_lo ? gvn->transform(new AddINode(index, new_add)) : index;
   }
   Node *new_cmp = (flip == 1)
-    ? new CmpUNode( new_add, range )
-    : new CmpUNode( range, new_add );
+    ? new CmpUNode(new_add, range)
+    : new CmpUNode(range, new_add);
   new_cmp = gvn->transform(new_cmp);
   // See if no need to adjust the existing check
-  if( new_cmp == cmp ) return;
+  if (new_cmp == cmp) return;
   // Else, adjust existing check
-  Node *new_bol = gvn->transform( new BoolNode( new_cmp, bol->as_Bool()->_test._test ) );
-  igvn->rehash_node_delayed( iff );
-  iff->set_req_X( 1, new_bol, igvn );
+  Node* new_bol = gvn->transform(new BoolNode(new_cmp, bol->as_Bool()->_test._test));
+  igvn->rehash_node_delayed(iff);
+  iff->set_req_X(1, new_bol, igvn);
+  // Loads and range check Cast nodes that are control dependent on this range check now depend on multiple dominating
+  // range checks. These control dependent nodes end up at the lowest/nearest dominating check in the graph. To ensure
+  // that these Loads/Casts do not float above any of the dominating checks (even when the lowest dominating check is
+  // later replaced by yet another dominating check), we need to pin them at the lowest dominating check.
+  proj->pin_array_loads(igvn);
```

Comment:
`pin_array_loads` suggests we only care about `Load`. But the comment suggests otherwise.

I would also appreciate if the comment said why there are now multiple dependencies. Actually, the problem is that we **would** have multiple dependency, but we only have one dependency input we can set, hence forgetting about the others. Pinning makes sure that there is no bypassing of dependencies, right?

Reply:
> Actually, the problem is that we **would** have multiple dependency, but we only have one dependency input we can set, hence forgetting about the others. Pinning makes sure that there is no bypassing of dependencies, right?

Right.